US3129 Name lists

katherine-boost opened the pull request:
Previously, they were known by their creation time, but lists now have a **name**.  

* Existing DB records have the name backfilled to (e.g.) `List created at 13 Sep 2018,  1:25:50 pm`
* The name field was then made non-nullable.
* Names must be ≥ 5 characters

<img width="379" alt="list-name-edit-1" src="https://user-images.githubusercontent.com/43123718/45605329-68bbdf80-ba90-11e8-8d59-930dc818a8fb.png">
Click the pen icon to get ...
<img width="579" alt="list-name-edit-2" src="https://user-images.githubusercontent.com/43123718/45605328-68234900-ba90-11e8-8c77-e4c80adfed8c.png">

However the main point of this branch was to do some **TESTING**!  

## Test notes
* Tests so far intentionally focus on `List` (`Item` is not covered)
* I've understood from the book [Everyday Rails Testing with RSpec](https://leanpub.com/everydayrailsrspec) that controller tests are considered to be low-value & it's better to test these things via integration tests. This makes sense - however, I wrote a couple just for the practice.
* I haven't used Factories yet since the `List` model is so simple. I'd like to introduce them in future though.
* Capybara selectors: I know it's preferred to use page text rather than CSS classes, but most of my actions are accomplished by clicking icons, I had to use those icon classes as selectors.

## Questions
* **Tests**: are they written according to good convention?
   * especially the feature tests.
   * I've created a helper file (`spec/features/helper.rb`) for reusable functions, is that how it's usually done?
   * one question re: Capybara selectors inline below
* **Two possible errors on one page**: On the individual list page, you could have an error from renaming the list, or from adding an item. I solved this by setting them as `flash[:list_notice]` and `flash[:item_notice]` respectively in the relevant controller actions. Is this a good way to solve that problem?

katherine-boost committed:
US3129 Add ability to name list + List tests

katherine-boost committed:
US3129 Make list title editable from list page

katherine-boost committed:
US3129 Add feature test for editing list title (from list page)

Including configuring Selenium/Chromedriver correctly.

katherine-boost committed:
Refactor to use vanilla JS and remove dependency on jQuery

katherine-boost committed:
US3129 Make List.name non-nullable; backfill existing nils.

Add feature test covering the scenario of a user trying to update the name to something invalid.

When editing the list's name, autofocus the <input>.
This is done via JavaScript: if passing autofocus: true, this causes an error on the client-side due the the form element being hidden on page load.

Delete dead tests, update others.

katherine-boost committed:
Restore Rails' requirement for rails-ujs (OOPS)

I deleted this thinking it was part of jQuery and/or a couple of other UI-related gems I uninstalled because I decided not to use them.
I was wrong. Never remove this requirement!!!

Thanks to Tim & Andy for helping me figure this one out!

katherine-boost commented on spec/features/lists_spec.rb:
```diff
@@ -0,0 +1,67 @@
+
+    scenario 'user deletes a list' do
+      list = find('li', text: another_list_name)
+      delete_link = list.all('a').last.click # TODO: fragile
```
My HTML output looks like this:

```html
<li class="row">
   <a class="columns small-10" href="/lists/42">Katherine&#39;s awesome list</a>
   <a data-confirm="Are you sure?" class="columns small-2" data-remote="true" rel="nofollow" data-method="delete" href="/lists/42">
      <i class="fi-trash"></i>
   </a>        
</li>
```

`list` is the full `<li>`.  
To get the delete link, I need to look with `list` and find the link containing the icon (but not the icon itself).  
Is there a way to do this with Capybara selectors that I'm not seeing, or do I simply need to use XPath?

mrGrazy replied:
So it's really common to add css classes that identify the link you want, and to use capybara's css selectors to find it.

katherine-boost replied:
I'd have to add a CSS class to that second link just for this test, though. That's kosher?

tim-boost replied:
You don't have to add the class to any css file, for the purpose of testing only, similar to how you  might add a class to an element so that you can select it using Jquery, it could be something like class='confirm-link' on the link, that's pretty common

tim-boost commented on app/assets/javascripts/application.js:
```diff
@@ -15,4 +15,22 @@
+    listTitle.classList.toggle('hidden')
+    listTitleForm.classList.toggle('hidden')
+
+    const input = document.querySelector('#list_name')
```
Note: Not that I imagine anyone will be using older browsers such as IE11, but just a note this will break on IE11 and any other browsers that don't support ES6 as const is part of ES6 ( without adding a polyfill to make it compatible with ES5 only browsers ).

katherine-boost replied:
Good point, it did cross my mind. This app is only for modern browsers 😄

mrGrazy commented on app/controllers/lists_controller.rb:
```diff
@@ -18,9 +18,28 @@ def edit
+    @list = List.find(params[:id])
+
+    @list.update(list_params)
+    @list.name = list_params[:name]
```
Line 37 should do this. Is there a reason why you need to assign it again?

katherine-boost replied:
Oh nice catch. No, I refactored and evidently forgot to delete that line. 🤦‍♂️

katherine-boost replied:
Fixed in e41970e

mrGrazy commented on app/controllers/lists_controller.rb:
```diff
@@ -18,9 +18,28 @@ def edit
+    @list.update(list_params)
+    @list.name = list_params[:name]
+
+    @list.save if @list.valid?
```
`.save` calls `.valid?` and returns false if the record is invalid

katherine-boost replied:
Right! Thanks! Fixed in e41970e

mrGrazy commented on app/helpers/lists_helper.rb:
```diff
@@ -1,5 +1,5 @@
-    list.updated_at.localtime.strftime('%e %b %Y, %l:%M:%S %P')
+  def format_time(timestamp)
+    timestamp.localtime.strftime('%e %b %Y, %l:%M:%S %P')
   end
```
So another common way this is done in rails is through the `l` helper method (short for 'localize') which is here https://guides.rubyonrails.org/i18n.html#adding-date-time-formats.

katherine-boost replied:
Excellent, I refactored to use this in e41970e

mrGrazy commented on app/views/lists/_errors.html.erb:
```diff
@@ -0,0 +1,6 @@
+  <strong><%= pluralize(errors.count, 'error') %>:</strong>
+  <% intersperse(errors, ", ").each do |msg| %>
+    <span><%= msg %></span>
+  <% end %>
```
My gut reaction to your intersperse method is that it is unusual. But I am struggling right now to suggest a better way to do what you're doing in this context. I'll get back to you ;).

tim-boost replied:
I was thinking similar, maybe I need to understand the purpose here but could it be possible to do this with no loop and errors.join(", ") in a single span since content in multiple spans would display inline anyway.

Or if there needs to be a span for each error, perhaps do errors.each_with_index do |msg, index| and then you could use unless index == errors.size - 1 to check whether or not you need a "," or a newline, break etc

katherine-boost replied:
Yep, I made this more complicated than it needs to be. 😆 I've changed to use `.join(", ")` in e41970e

katherine-boost committed:
Tweak view

mrGrazy commented on app/views/lists/show.html.erb:
```diff
@@ -43,6 +56,10 @@
+  <%= render 'errors', errors: flash[:item_notice] %>
+<% end %>
+
 <%= link_to lists_path, class: 'button' do %>
```
So it's more common to use flash messages for something like "List item created" or "We couldn't add your item because there were errors" rather than for the specific validation errors. We would just use `@list.errors` or `@item.errors` at this point. Does that make sense?

tim-boost replied:
Had a chat this morning, potentially could use Action Cable to broadcast the errors, we were unable to use @list.errors or @item.errors as the page is a 'single page app' with no form postbacks. I think this is something we can come back to.

katherine-boost replied:
I see your point, however I wasn't sure how to access the correct variables given that I've worked in a more SPA type style.  
e.g., I'm not using a `/new` page, and in my index page (`ListsController.index`) I only have access to `@lists`, not a new list `@list` from which I can pick up the errors.  

Tim suggested looking into ActionCable. I'll check it out, but it would be great to talk about this when you're back @mrGrazy .

mrGrazy commented on spec/features/helpers.rb:
```diff
@@ -0,0 +1,6 @@
+  def create_list(name)
+    fill_in 'Name', with: name
+    click_button 'Create new list'
+  end
```
I like what you've done here and it works. In some projects people add a 'support' folder in `/spec/`. However most projects also use the gem `factory_bot` which is all about generating models you need for your tests.

Even in integration tests you can still reach into your rails namespace to do things like create models `List.create` etc and also asset things in the database. While it's not quite as 'pure' in the sense of everything being done through the browser, it can make it much faster to do setup for your tests this way. It also allows you to set up models that might not be possible to create through the interface.

katherine-boost replied:
Thanks Andy. Hopefully I'll have time to write some more tests (and test `Item`), in which case I'll probably set up a factory. Good point on being able to use the ActionRecord methods!

mrGrazy reviewed (commented):
Hey I've just sprinkled lots of thoughts around here Katherine. I think what you've done is really good, I've just commented on things I think you might find interesting. Happy to discuss tomorrow if anything doesn't make sense.

katherine-boost committed:
Fix "disappearing event handlers" issue

After editing the list title (at /lists/:id), I was unable to update the title a second time - the onClick events seemed to have disappeared.

Andy pointed me to the solution. It's a matter of Turbolinks, which basically makes Rails more SPA-y.

See also: https://guides.rubyonrails.org/working_with_javascript_in_rails.html#turbolinks

katherine-boost committed:
US3129 Make list title editable from index page

In order to do this cleanly, I've pulled out the link title + form to make it editable out into a partial.

katherine-boost committed:
Fix-ups: i18n, simplify code, etc

• intersperse method can be achieved simply via .join
• x.save calls x.valid?, so it's not necessary to include that as an extra step
• set up localisation config and use the #l helper

katherine-boost committed:
Tests: use CSS class as selector
